### PR TITLE
feat: prCommitsPerRunLimit

### LINKED ISSUE #4279: rate limit per org (or renovate run)
**What would you like Renovate to be able to do?**
- I want a max num of PR creation per renovate run (not per repo)

**My problem**
- I have 150 micro services
- our jenkins based CICD can get swamped easily
- we want to automatically maintain the dependencies and the Dockerfiles with Renovate
- I use github repo autodiscovery and filter the repos (one per service, I have 150 to process)
- my problem is if Alpine change his image version - for example - our jenkins is swamped with PR validations jobs 

**Describe the solution you'd like**
I want a prHourlyLimitGlobal to set a maximum PR creation (or modification) per Renovate run and exit after the max is reached

**Describe alternatives you've considered**
- breakdown my repos list in smaller batches of 10-20 repos and schedule them every 2 hours but it s not very maintainable
- prHourlyLimit ? but it s per repo so for me it not useful apparently

**Additional context**

### PR DESCRIPTION
**Pull request by jsi-ka**
Closes #4279

### CODE REVIEW

**Comment by ViceIce** [comment]
@jsi-ka Why you don't reuse your old branch? You can simply force-push to update your pr.

**Comment by jsi-ka** [comment]
> @jsi-ka Why you don't reuse your old branch? You can simply force-push to update your pr.
Yeah it was my branch technically but we are several working on this :D

File: lib/workers/global/limits.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+const limitsToInit = ['prCommitsPerRunLimit'];
+const l = {};
+const v = {};
+
+export function init() {
+  let i;
+  for (i = 0; i < limitsToInit.length; i += 1) {
+    if (process.env[limitsToInit[i]]) {
+      setLimit(limitsToInit[i], parseInt(process.env[limitsToInit[i]], 10));
```
Comment (lhopki01) [inline lib/workers/global/limits.ts]: What about treating it the same way the `autodiscoverFilter` is treated as this too is a self hosted global option.

Comment (lhopki01) [inline lib/workers/global/limits.ts]: The default value for this if not set should not be 10.  It should be treated as unlimited.

File: lib/workers/global/index.js [hunk 2]
```diff
@@ -38,8 +39,14 @@ async function start() {
     global.trustLevel = config.trustLevel || 'low';
     delete config.trustLevel;
     detectRenovateVersion();
+    // init limits
+    limits.init();
```
Comment (ViceIce) [inline lib/workers/global/index.js]: Pass `config`
```suggestion
    limits.init(config);
```

File: lib/workers/global/limits.ts (hunk 1 above)
Comment (ViceIce) [reply lib/workers/global/limits.ts]: Here we should not use `process.env`. Use passed `config`

File: lib/workers/global/limits.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { logger } from '../../logger';
+
+const limitsToInit = ['prCommitsPerRunLimit'];
+const l = {};
+const v = {};
+
+export function init() {
```
Comment (ViceIce) [inline lib/workers/global/limits.ts]:
```suggestion
// TODO: fix type
export function init(config: Record<string,any>) {
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/global/limits.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import { logger } from '../../logger';
+
+const limitsToInit = ['prCommitsPerRunLimit'];
+const l = {};
```
Comment (ViceIce) [inline lib/workers/global/limits.ts]: Please add typeinfo, eg 
```suggestion
const l: Record<string,numnber> = {};
```

File: lib/workers/global/limits.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+const v = {};
```
Comment (ViceIce) [inline lib/workers/global/limits.ts]: Here too
```suggestion
const v: Record<string,numnber> = {};
```

File: lib/workers/global/limits.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+      logger.info(
+        `4279 feat - Limits.init ${limitsToInit[i]} environment variable is not set. Ignoring ${limitsToInit[i]}`
+      );
+    }
+  }
+}
+
+export function setLimit(name: string, value: number) {
+  logger.info(`4279 feat - Limits.setLimit h[${name}] = ${value}`);
```
Comment (ViceIce) [inline lib/workers/global/limits.ts]:
```suggestion
  logger.debug(`Limits.setLimit h[${name}] = ${value}`);
```

File: lib/workers/global/limits.ts (hunk 1 above)
Comment (jsi-ka) [reply lib/workers/global/limits.ts]: I was not sure about how to treat this global parameter. But I think you are right about treating the same way as `autodiscoverFilter`

File: lib/workers/global/limits.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+const v = {};
+
+export function init() {
+  let i;
+  for (i = 0; i < limitsToInit.length; i += 1) {
+    if (process.env[limitsToInit[i]]) {
+      setLimit(limitsToInit[i], parseInt(process.env[limitsToInit[i]], 10));
+      v[limitsToInit[i]] = 0;
+    } else {
+      logger.info(
```
Comment (ViceIce) [inline lib/workers/global/limits.ts]:
```suggestion
      logger.debug(
```

File: lib/workers/global/limits.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        `4279 feat - Limits.init ${limitsToInit[i]} environment variable is not set. Ignoring ${limitsToInit[i]}`
```
Comment (ViceIce) [inline lib/workers/global/limits.ts]:
```suggestion
        `Limits.init ${limitsToInit[i]} environment variable is not set. Ignoring ${limitsToInit[i]}`
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/global/limits.ts (hunk 1 above)
Comment (jsi-ka) [reply lib/workers/global/limits.ts]: I changed it so it used definitions.ts

**Review by viceice — APPROVED** [review]
@rarkins  What do you think?

**Review by viceice — CHANGES_REQUESTED** [review]
@jsi-ka `ERROR: renovate-schema.json needs updating. Run 'yarn create-json-schema' and commit.`

**Comment by jsi-ka** [comment]
for us it s ready to merge so I am removing the wip

**Review by viceice — APPROVED** [review]
LGTM

**Comment by rarkins** [comment]
This needs de-conflicting. website/docs now got moved to docs/usage. I attempted to edit myself but maintainer edit permissions were not granted on this fork

File: lib/workers/repository/process/write.js [hunk 9]
```diff
@@ -36,6 +38,14 @@ async function writeUpdates(config, packageFiles, allBranches) {
       return res;
     }
     prsRemaining -= res === 'pr-created' ? 1 : 0;
+    const changesList = ['pr-created', 'pr-edited', 'automerged'];
+    if (changesList.indexOf(res) > -1) {
```
Comment (rarkins) [inline lib/workers/repository/process/write.js]:
```suggestion
    if (changesList.includes(res)) {
```

File: lib/workers/global/limits.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  let result;
+  if (typeof v[name] !== 'undefined') {
+    result = l[name] - v[name];
+  } else {
+    result = undefined;
+  }
+  return result;
+}
+
+export function incrementLimit(name: string, value: number) {
```
Comment (rarkins) [inline lib/workers/global/limits.ts]:
```suggestion
export function incrementLimit(name: string, value: number = 1) {
```

File: lib/workers/repository/process/write.js [hunk 11]
```diff
@@ -36,6 +38,14 @@ async function writeUpdates(config, packageFiles, allBranches) {
...
+      limits.incrementLimit('prCommitsPerRunLimit', 1);
```
Comment (rarkins) [inline lib/workers/repository/process/write.js]:
```suggestion
      limits.incrementLimit('prCommitsPerRunLimit');
```

File: lib/workers/repository/process/write.js [hunk 12]
```diff
@@ -36,6 +38,14 @@ async function writeUpdates(config, packageFiles, allBranches) {
       return res;
     }
     prsRemaining -= res === 'pr-created' ? 1 : 0;
+    const changesList = ['pr-created', 'pr-edited', 'automerged'];
```
Comment (rarkins) [inline lib/workers/repository/process/write.js]: What about the case where a branch is created but the PR is not? Wouldn't we be better to implement this in the `commitFilesToBranch` function in `git/storage.ts`?

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
I can't tell why the linting is failing. Have you tried running `yarn lint-fix`?

**Comment by ViceIce** [comment]
https://circleci.com/gh/renovatebot/renovate/15007?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

image

File: lib/workers/global/limits.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  let result;
+  if (typeof v[name] !== 'undefined') {
+    result = l[name] - v[name];
+  } else {
+    result = undefined;
+  }
+  return result;
+}
+
+export function incrementLimit(name: string, value: number = 1) {
```
Comment (ViceIce) [inline lib/workers/global/limits.ts]:
```suggestion
export function incrementLimit(name: string, value = 1) {
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/write.js (hunk 12 above)
Comment (badaboum79) [reply lib/workers/repository/process/write.js]: You are right it makes more sense moving it to storage.ts. And it looks cleaner.

Comment (badaboum79) [reply lib/workers/repository/process/write.js]: done

**Comment by badaboum79** [comment]
all changes requested so far added. Thanks for the reviews

**Comment by ViceIce** [comment]
On Jenkins a pr creation triggers a build too. Is that counted too?

File: test/workers/global/limits.spec.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
+const { init } = require('../../../lib/workers/global/limits');
```
Comment (ViceIce) [inline test/workers/global/limits.spec.js]: New files should be written in typescript.

File: lib/workers/global/index.js [hunk 15]
```diff
@@ -39,9 +40,14 @@ async function start() {
     global.trustLevel = config.trustLevel || 'low';
     delete config.trustLevel;
     detectRenovateVersion();
+    limits.init(config);
     setEmojiConfig(config);
     // Iterate through repositories sequentially
     for (const repository of config.repositories) {
+      if (limits.getLimitRemaining('prCommitsPerRunLimit') <= 0) {
+        logger.info('Max commits created for this run. Exiting.');
```
Comment (rarkins) [inline lib/workers/global/index.js]:
```suggestion
        logger.info('Max commits created for this run. Skipping all remaining repositories.');
```

File: lib/config/definitions.ts [hunk 16]
```diff
@@ -447,6 +447,14 @@ const options: RenovateOptions[] = [
     type: 'string',
     default: null,
   },
+  {
+    name: 'prCommitsPerRunLimit',
+    description:
+      'Set a maximum number of commits per Renovate run. Default is no limit.',
+    stage: 'global',
+    type: 'integer',
+    default: null,
```
Comment (rarkins) [inline lib/config/definitions.ts]: Would using `null` for "no limit" be better or `0` for no limit?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/definitions.ts (hunk 16 above)
Comment (jsi-ka) [reply lib/config/definitions.ts]: I have changed it to zero to illustrate it makes no difference

File: lib/workers/global/limits.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
+import { logger } from '../../logger';
+
+const limitsToInit = ['prCommitsPerRunLimit'];
+const l: Record<string, number> = {};
+const v: Record<string, number> = {};
+
+export function init(config: Record<string, any>) {
+  let i;
+  for (i = 0; i < limitsToInit.length; i += 1) {
```
[the remark this answers is not in the document]
Comment (jsi-ka) [reply lib/workers/global/limits.ts]: done that

File: lib/platform/git/storage.ts [hunk 18]
```diff
@@ -366,6 +368,7 @@ export class Storage {
     await this._git!.checkout(this._config.baseBranch);
     await this._git!.merge(['--ff-only', branchName]);
     await this._git!.push('origin', this._config.baseBranch);
+    limits.incrementLimit('prCommitsPerRunLimit');
```
[the remark this answers is not in the document]
Comment (jsi-ka) [reply lib/platform/git/storage.ts]: I am not sure about this one... Could you elaborate please ?

**Comment by rarkins** [comment]
Thanks, @jsi-ka !

**Pull request merged**